foreshadow/column_sharer.py

```diff
@@ -0,0 +1,179 @@
+"""Cache utility for foreshadow pipeline workflow to share data."""
+from collections import MutableMapping, defaultdict
+
+
+class ColumnSharer(MutableMapping):
+    """main cache-class to be used as single-instance to share data."""
+
+    def __init__(self, *args, **kwargs):
+        super(ColumnSharer, self).__init__(*args, **kwargs)
+        self.store = defaultdict(lambda: {})  # will have a nested dict for
+        # every key, which holds {column: key-column info}
+        acceptable_keys = {"intent": True, "domain": True, "metastat": True}
+        for key in acceptable_keys:
+            self.store[key] = {}
+        self.__acceptable_keys = defaultdict(lambda: False, acceptable_keys)
+        self.__registered_keys = defaultdict(lambda: False)
+
+    def __getitem__(self, key_list):
+        """Override getitem to support multi key accessing simultaneously.
+
+        Args:
+            key_list: list of keys passed
+
+        Returns:
+            [key] or [key][column] from internal dict.
+
+        Raises:
+              e: exception raised by dict on invalid access. Likely
+                  a KeyError.
+
+        """
+        # first, get the item from the dict
+        key, column = self._convert_key(key_list)
+        self.check_key(key)
+        key_dict = self.store[key]
+        if column is not None:  # then get the column if requested
+            try:
+                return key_dict[column]
+            except Exception as e:
+                raise e
+        return key_dict  # otherwise return all the columns
+
+    def __setitem__(self, key_list, value):
+        """Enable multi key setting simultaneously.
+
+        Args:
+            key_list: list of keys passed
+            value: value to set, dict if just the key and value if key, column
+
+        """
+        key, column = self._convert_key(key_list)
+        self.check_key(key)
+        if column is None:  # setting the value for the entire key
+            self.store[key] = value
+        else:  # setting a particular column's value for a given key.
+            self.store[key][column] = value
+
+    def check_key(self, key):
+        """Check they passed key to see if it is a valid key.
+
+        Args:
+            key: the key passed to this object.
+
+        Raises:
+            KeyError: if not a valid key (predefined or registered)
+
+        """
+        if not self.__acceptable_keys[key] and not self.__registered_keys[key]:
+            raise KeyError(
+                "key {} is not a valid key. Please register it "
+                "first".format(key)
+            )
```

Comment:
Silly question, why are are the default keys not "registered" in acceptable keys, would this error out?